vault/rollback.go

```diff
@@ -132,24 +144,62 @@ func (m *RollbackManager) triggerRollbacks() {
 		}
 		fullPath := e.namespace.Path + path
 
-		m.inflightLock.RLock()
-		_, ok := m.inflight[fullPath]
-		m.inflightLock.RUnlock()
-		if !ok {
-			m.startRollback(ctx, fullPath, true)
-		}
+		// Start a rollback if necessary
+		m.startOrLookupRollback(ctx, fullPath, true)
 	}
 }
 
 // startRollback is used to start an async rollback attempt.
 // This must be called with the inflightLock held.
-func (m *RollbackManager) startRollback(ctx context.Context, fullPath string, grabStatelock bool) *rollbackState {
-	rs := &rollbackState{}
-	rs.Add(1)
-	m.inflightAll.Add(1)
+func (m *RollbackManager) startOrLookupRollback(ctx context.Context, fullPath string, grabStatelock bool) *rollbackState {
+	rs := &rollbackState{
+		rollbackFunc: func(ctx context.Context) error {
+			ns, err := namespace.FromContext(ctx)
+			if err != nil {
+				return err
+			}
+			if ns == nil {
+				return namespace.ErrNoNamespace
+			}
+
+			// Invoke a RollbackOperation
+			req := &logical.Request{
+				Operation: logical.RollbackOperation,
+				Path:      ns.TrimmedPath(fullPath),
+			}
+
+			var cancelFunc context.CancelFunc
+			ctx, cancelFunc = context.WithTimeout(ctx, DefaultMaxRequestDuration)
+			_, err = m.router.Route(ctx, req)
+			cancelFunc()
+
+			// If the error is an unsupported operation, then it doesn't
+			// matter, the backend doesn't support it.
+			if err == logical.ErrUnsupportedOperation {
+				err = nil
+			}
+			// If we failed due to read-only storage, we can't do anything; ignore
+			if err != nil && strings.Contains(err.Error(), logical.ErrReadOnly.Error()) {
+				err = nil
+			}
+			if err != nil {
+				m.logger.Error("error rolling back", "path", fullPath, "error", err)
+			}
+			return nil
+		},
+	}
+
 	m.inflightLock.Lock()
+	defer m.inflightLock.Unlock()
+	rsInflight, ok := m.inflight[fullPath]
+	if ok {
+		return rsInflight
+	}
+
```

Comment:
Why not move the creation of rs down here, so that we don't generate garbage needlessly?

Reply:
Hah, I made just this comment but up above.